src/protocol/context/malicious.rs

```diff
@@ -58,17 +71,47 @@ impl<'a, F: Field> MaliciousContext<'a, F> {
     pub(super) fn new<S: Substep + ?Sized>(
         source: &SemiHonestContext<'a>,
         malicious_step: &S,
-        upgrade_ctx: SemiHonestContext<'a>,
         acc: MaliciousValidatorAccumulator<F>,
         r_share: Replicated<F>,
     ) -> Self {
         Self {
-            inner: ContextInner::new(upgrade_ctx, acc, r_share),
+            inner: ContextInner::new(source, acc, r_share),
             step: source.step().narrow(malicious_step),
             total_records: TotalRecords::Unspecified,
         }
     }
 
+    /// TODO: This is not fast, but we can't just `reinterpret_cast` here.
+    fn as_semi_honest(&self) -> SemiHonestContext<'a> {
+        SemiHonestContext::new_complete(
+            self.inner.prss,
+            self.inner.gateway,
+            self.step.clone(),
+            self.total_records,
+        )
+    }
+
+    async fn upgrade_one(
+        &self,
+        record_id: RecordId,
+        x: Replicated<F>,
+        zeros_at: ZeroPositions,
+    ) -> Result<MaliciousReplicated<F>, Error> {
+        let rx = x
+            .multiply_sparse(
+                &self.inner.r_share,
+                self.as_semi_honest(),
+                record_id,
+                (zeros_at, ZeroPositions::Pvvv),
+            )
+            .await?;
+        let m = MaliciousReplicated::new(x, rx);
+        let narrowed = self.narrow(&RandomnessForValidation);
```

Comment:
I also don't want to narrow the entire context in order to get this value back, but the whole instrumentation layer makes that more awkward than I'd like.